### PR TITLE
3DFileViewer+LibGfx: Use unveil & FileSystemAccessServer

### PR DESCRIPTION
**Pull request by timmot**
* **LibGfx: Add methods to load image from existing fd**

* **3DFileViewer: Use unveil and FileSystemAccessServer**
   This will restrict 3DFileViewer's access to the file system.

   3DFileViewer loads a texture based on the path of the model loaded, this
    will request access to the texture file before loading.

### CODE REVIEW

File: Userland/Applications/3DFileViewer/main.cpp [hunk 1]
```diff
@@ -182,26 +185,43 @@ void GLContextWidget::timer_event(Core::TimerEvent&)
     m_cycles++;
 }
 
-bool GLContextWidget::load(const String& filename)
+bool GLContextWidget::load_path(String const& filename)
 {
     auto file = Core::File::construct(filename);
 
-    if (!file->filename().ends_with(".obj")) {
```
Comment (timmot) [inline Userland/Applications/3DFileViewer/main.cpp:189]: Add this back.

**Comment by timmot** [comment]
> Let's find a way to do this at a higher level, so that decoder plugins don't need to know anything about file descriptors. :)

I can see a way forward by stripping the decoder plugins to `load_xxx_from_memory` and moving the `MappedFile` creation into Bitmap itself.

Currently this would remove the filename from the mmap name, but we could add that back by adding a `name` parameter to the `load_xxx_from_memory` functions.

Does this sound reasonable?

**Comment by awesomekling** [comment]
> > Let's find a way to do this at a higher level, so that decoder plugins don't need to know anything about file descriptors. :)
> 
> I can see a way forward by stripping the decoder plugins to `load_xxx_from_memory` and moving the `MappedFile` creation into Bitmap itself.
> 
> Currently this would remove the filename from the mmap name, but we could add that back by adding a `name` parameter to the `load_xxx_from_memory` functions.
> 
> Does this sound reasonable?

Yes, that sounds very reasonable! :)

File: Userland/Applications/3DFileViewer/CMakeLists.txt [hunk 2]
```diff
@@ -2,6 +2,7 @@ serenity_component(
     3DFileViewer
     RECOMMENDED
     TARGETS 3DFileViewer
+    DEPENDS FileSystemAccessServer
```
Comment (mustafaquraish) [inline Userland/Applications/3DFileViewer/CMakeLists.txt:5]: Is this needed? I'm not too hot on CMake, but I don't think the other applications have this.

File: Userland/Applications/3DFileViewer/main.cpp [hunk 3]
```diff
@@ -226,8 +246,21 @@ bool GLContextWidget::load(const String& filename)
     builder.append(filename.split('.').at(0));
     builder.append(".bmp");
 
+    String texture_path = Core::File::absolute_path(builder.string_view());
+
     // Attempt to open the texture file from disk
-    auto texture_image = Gfx::Bitmap::try_load_from_file(builder.string_view());
+    RefPtr<Gfx::Bitmap> texture_image;
+    if (Core::File::exists(texture_path)) {
```
Comment (mustafaquraish) [inline Userland/Applications/3DFileViewer/main.cpp:253]: Why try to open this using `Core::File` if we're using FSAC instead? This just requires us to have different methods to load from name / fd+name, etc.

File: Userland/Applications/3DFileViewer/main.cpp [hunk 4]
```diff
@@ -249,11 +282,36 @@ int main(int argc, char** argv)
...
         perror("pledge");
         return 1;
     }
 
+    if (unveil("/tmp/portal/filesystemaccess", "rw") < 0) {
+        perror("unveil");
+        return 1;
+    }
+
+    if (unveil("/home/anon/Documents/3D Models/teapot.obj", "r") < 0) {
```
Comment (mustafaquraish) [inline Userland/Applications/3DFileViewer/main.cpp:295]: Seems weird to unveil these explicitly if we're using FSAC.

File: Userland/Libraries/LibGfx/JPGLoader.cpp [hunk 5]
```diff
@@ -1238,17 +1237,14 @@ RefPtr<Gfx::Bitmap> load_jpg(String const& path)
...
         return nullptr;
-    auto bitmap = load_jpg_impl((const u8*)file_or_error.value()->data(), file_or_error.value()->size());
-    if (bitmap)
-        bitmap->set_mmap_name(String::formatted("Gfx::Bitmap [{}] - Decoded JPG: {}", bitmap->size(), LexicalPath::canonicalized_path(path)));
-    return bitmap;
+    return load_jpg_from_memory((const u8*)file_or_error.value()->data(), file_or_error.value()->size(), path);
 }
 
-RefPtr<Gfx::Bitmap> load_jpg_from_memory(const u8* data, size_t length)
+RefPtr<Gfx::Bitmap> load_jpg_from_memory(const u8* data, size_t length, String const& mmap_name)
```
Comment (mustafaquraish) [inline Userland/Libraries/LibGfx/JPGLoader.cpp:1243]: if you're touching this anyway, turn `const u8*` to `u8 const*` (and other lines you've already modified)

File: Userland/Applications/3DFileViewer/CMakeLists.txt (hunk 2 above)
Comment (timmot) [reply Userland/Applications/3DFileViewer/CMakeLists.txt:5]: I believe this is required for the component system to work.

File: Userland/Applications/3DFileViewer/main.cpp (hunk 4 above)
Comment (timmot) [reply Userland/Applications/3DFileViewer/main.cpp:295]: This is absolutely fine for sandboxing. It is nice to reduce the amount of unveils but these are demo files which are opened every time the application is opened.

File: Userland/Applications/3DFileViewer/main.cpp (hunk 3 above)
Comment (timmot) [reply Userland/Applications/3DFileViewer/main.cpp:253]: This duplication exists to let the application load the teapot on start-up in conjunction with the unveils. I think this makes sense unless the teapot on start-up is removed.

File: Userland/Libraries/LibGfx/JPGLoader.cpp (hunk 5 above)
Comment (timmot) [reply Userland/Libraries/LibGfx/JPGLoader.cpp]: Sounds good, done for all loaders.

**Comment by timmot** [comment]
> I can see a way forward by stripping the decoder plugins to `load_xxx_from_memory` and moving the `MappedFile` creation into Bitmap itself.
> Currently this would remove the filename from the mmap name, but we could add that back by adding a `name` parameter to the `load_xxx_from_memory` functions.

These changes have been made :)

**Pull request merged**